File: server/app/routes/blue.py
from __future__ import annotations

from datetime import UTC, datetime
from pathlib import Path
from typing import Any

from fastapi import APIRouter, HTTPException

from app.agents.secure_gemini import run_secure_gemini
from app.defense.delimiters import wrap_untrusted
from app.defense.detectors import detect_injection_signals
from app.pdf.extract import extract_text, highlight_injections
from app.pdf.sanitize import sanitize_resume_text
from app.store.runs import get_run, push_event, update_run

router = APIRouter(prefix="/api/blue", tags=["blue"])


@router.post("/defend")
async def defend(body: dict[str, Any]):
    run_id = body.get("run_id")
    if not run_id:
        raise HTTPException(400, "run_id required")
    run = get_run(run_id)
    if not run:
        raise HTTPException(404, "run not found")

    pdf = Path(run["poisonedPdfPath"]).read_bytes()
    extracted = extract_text(pdf)
    sanitize = sanitize_resume_text(extracted["text"])
    detections = detect_injection_signals(extracted["text"])
    delimited = wrap_untrusted(sanitize["cleanedText"])
    agent = await run_secure_gemini(delimited, payload_count=run.get("injectionCount"))

    def mutate(r: dict[str, Any]) -> None:
        r["extractedText"] = extracted["text"]
        r["highlights"] = highlight_injections(extracted["text"], r["injected"])
        r["blue"] = {
            "completedAt": datetime.now(UTC).isoformat(),
            "sanitize": sanitize,
            "detections": detections,
            "delimitedPreview": delimited[:2000],
            "agent": agent,
            "alerts": agent["alerts"],
        }
        push_event(
            r,
            "blue_defend_complete",
            {
                "blockedTools": len(agent["blockedTools"]),
                "alerts": len(agent["alerts"]),
                "findings": len(sanitize["findings"]),
            },
        )

    updated = update_run(run_id, mutate)

    return {
        "run_id": run_id,
        "mode": "blue",
        "extractedText": extracted["text"],
        "highlights": updated.get("highlights"),
        "sanitize": sanitize,
        "detections": detections,
        "delimitedPreview": delimited[:2000],
        "thoughts": agent["thoughts"],
        "modelText": agent["text"],
        "toolCalls": agent["toolCalls"],
        "blockedTools": agent["blockedTools"],
        "mail": agent["mailResults"],
        "alerts": agent["alerts"],
        "timeline": updated.get("timeline"),
    }

File: server/src/routes/blue.ts
import { Router } from "express";
import { extractText, highlightInjections } from "../pdf/extract.js";
import { sanitizeResumeText } from "../pdf/sanitize.js";
import { wrapUntrusted } from "../defense/delimiters.js";
import { detectInjectionSignals } from "../defense/detectors.js";
import { runSecureGemini } from "../agents/secureGemini.js";
import { buildCleanedResumePdf } from "../pdf/inject.js";
import {
  getRun,
  updateRun,
  pushEvent,
  storePdf,
  loadPdf,
  pdfKeyPoisoned,
  pdfKeyCleaned,
  pdfKeyOriginal,
} from "../store/runs.js";

export const blueRouter = Router();

blueRouter.post("/defend", async (req, res, next) => {
  try {
    const runId = req.body.run_id as string;
    if (!runId) {
      res.status(400).json({ error: "run_id required" });
      return;
    }
    const run = await getRun(runId);
    if (!run) {
      res.status(404).json({ error: "run not found" });
      return;
    }

    // Load PDF from in-memory store
    const pdf = loadPdf(pdfKeyPoisoned(runId));
    if (!pdf) {
      res.status(404).json({ error: "PDF not found — run may have expired (serverless ephemeral memory)" });
      return;
    }

    const extracted = await extractText(pdf);
    const sanitize = sanitizeResumeText(extracted.text);
    const detections = detectInjectionSignals(extracted.text);
    const delimited = wrapUntrusted(sanitize.cleanedText);
    const agent = await runSecureGemini(delimited, { payloadCount: run.injectionCount });

    // If the model stayed quiet but sanitize found injections, still raise an admin alert.
    if (agent.alerts.length === 0 && sanitize.suspectedInjectionCount > 0) {
      const { interceptSendEmail } = await import("../mail/interceptor.js");
      const probe = await interceptSendEmail(
        {
          to: "attacker@evil.example",
          subject: "Blocked probe — injection signatures in resume",
          body: `Sanitizer findings: ${sanitize.suspectedInjectionCount}. Excerpt: ${sanitize.findings[0]?.excerpt || ""}`,
        },
        { payloadCount: run.injectionCount, toolArgs: { source: "sanitize_detector" } },
      );
      if (probe.alert) agent.alerts.push(probe.alert);
    }

    // Build a clean PDF (injections stripped) and store in-memory
    const cleanedBytes = await buildCleanedResumePdf(sanitize.cleanedText);
    storePdf(pdfKeyCleaned(runId), Buffer.from(cleanedBytes));

    const cleanPdfUrl = `/api/files/${runId}-cleaned.pdf`;
    const originalPdfUrl = loadPdf(pdfKeyOriginal(runId))
      ? `/api/files/${runId}-original.pdf`
      : undefined;

    const updated = await updateRun(runId, (r) => {
      r.extractedText = extracted.text;
      r.highlights = highlightInjections(extracted.text, r.injected);
      r.blue = {
        completedAt: new Date().toISOString(),
        sanitize,
        detections,
        delimitedPreview: delimited.slice(0, 2000),
        agent,
        alerts: agent.alerts,
        cleanPdfUrl,
        originalPdfUrl,
      };
      pushEvent(r, "blue_defend_complete", {
        blockedTools: agent.blockedTools.length,
        alerts: agent.alerts.length,
        findings: sanitize.findings.length,
        cleanPdfUrl,
      });
    });

    res.json({
      run_id: runId,
      mode: "blue",
      extractedText: extracted.text,
      highlights: updated.highlights,
      sanitize,
      detections,
      delimitedPreview: delimited.slice(0, 2000),
      thoughts: agent.thoughts,
      modelText: agent.text,
      toolCalls: agent.toolCalls,
      blockedTools: agent.blockedTools,
      mail: agent.mailResults,
      alerts: agent.alerts,
      timeline: updated.timeline,
      cleanPdfUrl,
      originalPdfUrl,
    });
  } catch (err) {
    next(err);
  }
});
